soroban-env-host/src/host/crypto.rs

```diff
@@ -238,3 +241,47 @@ pub(crate) fn chacha20_fill_bytes(
     rng.fill_bytes(dest);
     Ok(())
 }
+
+// It is possible that a user-provided PRNG seed (either in a test or, more
+// worryingly, in a production environment) is biased: it might be all zero, or
+// all copies of a single byte, or otherwise statistically unlike a uniformly
+// random bitstream with roughly 50-50 zero and one bits.
+//
+// Unfortunately the security properties of the stream cipher ChaCha used in the
+// PRNG (being "indistinguishable from uniform random") are based on the
+// assumption of an _unbiased_ seed.
+//
+// So we run any seed through HMAC-SHA256 here, with a constant uniform random
+// salt, as an unbiasing step (this is the "randomness-extractor" phase of HKDF,
+// which is the only part relevant to our needs, we don't need multiple keys).
+
+pub(crate) fn unbias_prng_seed(
+    seed: &[u8; SEED_BYTES as usize],
+    budget: impl AsBudget,
+) -> Result<[u8; SEED_BYTES as usize], HostError> {
+    tracy_span!("unbias_prng_seed");
+
+    // Salt is fixed and must not be changed; it is effectively "part of the
+    // protocol" and must be the same for all implementations.
+    //
+    // Note: salt is a "public random value", intended to be statistically
+    // similar to a 32-byte draw on /dev/random but done in a transparent and
+    // reproducible way. In this case we use the Stellar Public Network ID,
+    // `sha256("Public Global Stellar Network ; September 2015")`.
+    //
+    // This number as a bitstring has 137 zeroes and 119 ones, which is within
+    // the range we get when taking 32-byte samples from /dev/random (feel free
+    // to check this yourself).
+
+    const SALT: [u8; 32] = hex!("7ac33997544e3175d266bd022439b22cdb16508c01163f26e5cb2a3e1045a979");
+
+    // Running HMAC will will run SHA256 2 times.
+    budget
+        .as_budget()
+        .charge(ContractCostType::ComputeSha256Hash, Some(2))?;
+
+    let mut hmac = Hmac::<Sha256>::new_from_slice(&SALT)
+        .map_err(|_| Error::from_type_and_code(ScErrorType::Context, ScErrorCode::InternalError))?;
```

Comment:
nit: `Hmac::new_from_slice` is infallible. This function returns a `Result` because the trait allows for the possibility that a key could be a bad length, but for HMAC any length is valid. So in principle you could `.expect("valid HMAC key length")` safely here. Not sure if that jibes with the local style, though :)

Reply:
yeah we're being extra paranoid about panics and turning them all into "internal errors" to minimize the odds of a stray unwind due to a mistake.